### ISSUE TITLE
'through' association not working as expected

### ISSUE THREAD

**Issue by daVinciWanaB** [cakephp/cakephp#6475]
When I use 'joinTable', my association works, but has other problems (different discussion).  So, I re-read the book, and found out that I should have been using the 'through' option, but it's not working as expected.

    // VideosTable
		$this->belongsToMany('Tags', [
			'through' => 'Tagged',
			'targetForeignKey' => 'foreign_key',
			'conditions' => [
				'Tagged.foreign_key_table' => 'videos'
			]
		]);

But, when I try to do a `get()` in my VideosController:

		$video = $this->Videos->get($id, ['contain' => ['Tags']]);

I get an error page:
Error: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'Tagged.video_id' in 'where clause'

It doesn't show the standard left rail with stacktrace and debugkit doesn't load, so I've been hard-pressed to figure out what's going on or why/how it could still be looking for video_id even though I believe I'm specifying it shouldn't be.

I'm a Cake 2 veteren-ish, but am new to Cake3.  Upon many attempts to get help on this, and trying every seemingly possible combination of associations, key settings, options...etc, it just doesn't seem to work.  I tried walking through it w/ joze_zap (who obviously knows a lot more than me) and it was recommended I put in a ticket.

Please let me know if there is ANY other information I can provide.  I realize this still may be MY issue, but it's a very lightweight app so far (not a lot of moving parts or customizations), and I just can't seem to figure out what's going on.

Files:
Admin/VideosController:  https://gist.github.com/daVinciWanaB/e33a4ce00844d63fcd6c
VideosTable:  https://gist.github.com/daVinciWanaB/ff4a48ce99108b64f761
TaggedTable:  https://gist.github.com/daVinciWanaB/f0b2abb875b31a1c236c
TagsTable:  https://gist.github.com/daVinciWanaB/72f8f2d6c974c79e2da4

**Comment by markstory** [comment]
What does the schema for the three tables look like? I wouldn't need all the columns, just the ones relevant to the associations.

**Comment by daVinciWanaB** [comment]
CREATE TABLE IF NOT EXISTS `tagged` (
  `id` char(36) NOT NULL,
  `tag_id` char(36) NOT NULL,
  `foreign_key` char(36) DEFAULT NULL,
  `foreign_key_table` varchar(60) NOT NULL,
  `created` datetime NOT NULL,
  `modified` datetime NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=MyISAM DEFAULT CHARSET=latin1;

CREATE TABLE IF NOT EXISTS `tags` (
  `id` char(36) NOT NULL,
  `namespace` varchar(255) DEFAULT NULL,
  `slug` varchar(255) NOT NULL,
  `label` varchar(255) NOT NULL,
  `counter` int(11) NOT NULL DEFAULT '0',
  `created` datetime DEFAULT NULL,
  `modified` datetime DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=MyISAM DEFAULT CHARSET=latin1;

CREATE TABLE IF NOT EXISTS `videos` (
  `id` char(36) NOT NULL DEFAULT '',
  `title` varchar(200) DEFAULT NULL,
  `youtube_id` varchar(20) DEFAULT NULL,
  `length` char(5) DEFAULT NULL,
  `yt_watch_url` varchar(200) DEFAULT NULL,
  `yt_description` varchar(1500) DEFAULT NULL,
  `description` varchar(500) DEFAULT NULL,
  `thumbnail` varchar(200) DEFAULT NULL,
  `slug` varchar(120) DEFAULT NULL,
  `model` varchar(30) DEFAULT NULL,
  `foreign_key` char(36) DEFAULT NULL,
  `rating_count` int(11) DEFAULT NULL,
  `created` datetime DEFAULT NULL,
  `modified` datetime DEFAULT NULL,
  `approved` tinyint(1) DEFAULT '0',
  `rating` int(1) DEFAULT NULL,
  `featured` tinyint(1) NOT NULL DEFAULT '0',
  `tag_count` int(11) NOT NULL DEFAULT '0',
  `yt_title` varchar(200) DEFAULT NULL,
  `yt_thumbnail_url` varchar(200) DEFAULT NULL,
  `yt_length` varchar(8) DEFAULT NULL,
  `yt_rating` float(10,0) DEFAULT NULL,
  `yt_comments_url` varchar(200) DEFAULT NULL,
  `yt_comments_count` int(10) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=MyISAM DEFAULT CHARSET=latin1;

**Comment by ADmad** [comment]
In your belongsToMany association instead of `'targetForeignKey' => 'foreign_key'`  it should be `'foreignKey' => 'foreign_key'`.

**Comment by daVinciWanaB** [comment]
ADmad: doing that causes another error:  Error: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'Tagged.foreign_key_table' in 'where clause'

(Basically, this is what I've been doing - changing all the options to everything I could think, then getting a new error, changing it, getting a new error...etc in a vicious cycle)

And according to the Cake book the 'foreignKey' is to specify "the name of the foreign key found in the current model".  I suppose I'm probably just misinterpreting that, but in this case there IS no foreign key in the current model (unless it means the Videos.id field).  I do see further on where it says the default is the current model + "_id", so that makes sense as to how it was looking for 'video_id'.

**Comment by ADmad** [comment]
I guess "the name of the foreign key found in the current model" should probably say "the name of the foreign key **for** current model"

**Comment by daVinciWanaB** [comment]
ADmad: that would make more sense, you're correct.  I've also removed the conditions from the VideosTable belongsToMany association and put them in the TaggedTable association, and it seems to be working correctly-enough for now.  Thanks for both of your time.  I assume the rest of the issues I'm having are now related to something else, and all of this was due to my lack of understanding.

Side note: One of the things I couldn't figure out (that joze_zap was surprised about) was that the stack trace and DebugKit didn't load at all when these errors were thrown. Again, that's likely another issue alltogether, but - just an FYI I guess.

Thanks again for your time.
